build-system/babel-plugins/babel-plugin-transform-function-declarations/index.js

```diff
@@ -14,119 +14,198 @@
-    referencesAreOnlyCallExpressions(path) {
-      const {name} = path.get('id').node;
-      const binding = path.scope.getBinding(name);
-      return !binding.referencePaths.every((p) =>
-        p.parentPath.isCallExpression({callee: p.node})
-      );
-    },
-  };
+  /**
+   * If the FunctionDeclaration contains a ThisExpression, converting from a FunctionDeclaration to a
+   * VariableDeclaration => VariableDeclarator => ArrowFunctionExpression isn't necessarily valid.
+   * @param {*} path
+   * @return {boolean}
+   */
+  function containsThisExpression(path) {
+    let containsThis = false;
+    path.traverse({
+      ThisExpression() {
+        containsThis = true;
+        path.stop();
+      },
+    });
+    return containsThis;
+  }
 
-  // Making an existing FunctionDeclaration a let VariableDeclaration is safe.
-  // The name was already reserved for the FunctionDeclaration.
-  function createVariableDeclaration(path) {
-    const {params, body, async, id} = t.cloneNode(path.node);
-    const arrowFunction = t.arrowFunctionExpression(
-      params,
-      body.body[0].argument,
-      async
+  /**
+   * If the FunctionDeclaration identifier is used a manner besides a CallExpression, bail.
+   * @param {*} path
+   * @param {string} name
+   * @return {boolean}
+   */
+  function referencesAreOnlyCallExpressions(path, name) {
+    const binding = path.scope.getBinding(name);
+    return !binding.referencePaths.every((p) =>
+      p.parentPath.isCallExpression({callee: p.node})
     );
-    const declarations = t.variableDeclarator(
-      t.identifier(id.name),
-      arrowFunction
-    );
-    const declaration = t.variableDeclaration('let', [declarations]);
+  }
+
+  function createArrowFunctionExpression(node) {
+    const {params, body, async} = t.cloneNode(node);
+    return t.arrowFunctionExpression(params, body.body[0].argument, async);
+  }
+
+  function createVariableDeclarator(path, name) {
+    const arrowFunction = createArrowFunctionExpression(path.node);
+    return t.variableDeclarator(t.identifier(name), arrowFunction);
+  }
+
+  function createVariableDeclaration(path) {
+    const declarator = createVariableDeclarator(path, path.node.id.name);
+    const declaration = t.variableDeclaration('let', [declarator]);
 
     t.inherits(declaration, path.node);
     return declaration;
   }
 
+  const DECLARATION_BAIL_OUT_CONDITIONS = [
+    isNotFunction,
+    isGenerator,
+    isNotSingleReturnStatement,
+    containsArgumentsUsage,
+    containsThisExpression,
+    (path) => referencesAreOnlyCallExpressions(path, path.get('id').node.name),
+  ];
+
+  // If CallExpression names should be exhempt from arrow conversion, denote them here.
+  const EXHEMPTED_EXPRESSION_NAME_REGEXS = [/registerService/];
+
+  const EXPRESSION_BAIL_OUT_CONDITIONS = [
+    isNotFunction,
+    isGenerator,
+    isNotSingleReturnStatement,
+    containsArgumentsUsage,
+    containsThisExpression,
+    (path) => {
```

Comment:
why is this special cased as an anon (no named) condition?

Reply:
Both `Declarations` and `Expressions` have this final condition calling into `referencesAreOnlyCallExpressions`, but the `Expression` path adds in verification that the name of a method isn't part of the exemption list (`EXHEMPTED_EXPRESSION_NAME_REGEXS`).

I could separate this check into its own bail out condition if it would provide additional clarity.